Allow "auto"  layout args for the `create_compute_pipeline`

dskkato opened the pull request:
I attempted the implementation to allow "auto" for the layout in `create_compute_pipeline`, but to be honest, I'm not entirely clear on how the behavior of `wgpuComputePipelineGetBindGroupLayout` is defined for the pipeline created with ffi.NULL as the layout args. As implemented in this PR, it seems to work correctly only when the `group_id` is 0, and otherwise, it crashes.

However, I can confirm that applying this patch enables the same syntax as the wgpu compute shader example as below.

https://github.com/gfx-rs/wgpu/blob/fd53ea90e675d94f1d79a1c3c44b2f356cecd9c5/examples/hello-compute/src/main.rs#L106-L123

I'm not sure if this patch can be merged, but I'm documenting the necessary steps here as a memo for the work required to accept "auto." If it seems like progress can be made, I'll proceed with the necessary tasks.

almarklein reviewed (commented):
Thanks! This looks good, and you also added a test 👍   I only have a small comment to cover the else clause in the triage on `layout`.

> As implemented in this PR, it seems to work correctly only when the group_id is 0, and otherwise, it crashes.

I'll have a look too.

almarklein commented:
>> As implemented in this PR, it seems to work correctly only when the group_id is 0, and otherwise, it crashes.
>
> I'll have a look too.

It *nearly* works, but there is no way (that I know of) to know how many bind groups there are.

@rajveermalviya this PR implements auto-layouts, it uses `wgpuComputePipelineGetBindGroupLayout()` to get the automatically generated layout objects, but how should we know how many there are? I'm not seeing a way to get this from the shader module object. Does webgpu.h perhaps need something like `wgpuComputePipelineGetBindGroupLayoutCount()`?

dskkato commented:
I've added two commits on my end.

The first one is related to the implementation of `get_bind_group_layout`. The previous implementation assumed that the layout is provided when the pipeline is constructed, but in the `'auto'` layout mode, access is only possible after pipeline construction. Therefore, I modified `get_bind_group_layout` to call `wgpuComputePipelineGetBindGroupLayout` every time. This results in the instantiation of a new `GPUBindGroupLayout` instance each time, but since it aligns with the WebGPU spec, I believe this implementation is acceptable. One concern is the current implementation of `wgpuComputePipelineGetBindGroupLayout` in wgpu-native, which immediately aborts if an out-of-range index is accessed. It would be beneficial if we could either retrieve the number of bind group layouts or handle errors at the caller side.

https://www.w3.org/TR/webgpu/#dom-gpupipelinebase-getbindgrouplayout

The second commit removes the unused _bindings similar to _layouts. Since this deletion could be done without modifying the tests, I think it should be fine.

Let me know if you have any comments or concerns.

almarklein commented on wgpu/backends/wgpu_native/_api.py:
```diff
@@ -1784,7 +1791,14 @@ def _destroy(self):
+        Note that current wgpu-native aborts immediately if the index is out of range.
+        """
+        # H: WGPUBindGroupLayout f(WGPUComputePipeline computePipeline, uint32_t groupIndex)
+        layout_id = libf.wgpuComputePipelineGetBindGroupLayout(self._internal, index)
```
Bonus points to cover both cases in the tests 😉

dskkato replied:
Thank you for pointing that out. When I implemented this, I thought I was implementing it in ComputePipeline, but since this is in PipelineBase, as you mentioned, checks are indeed necessary. I will check if it can be implemented.

almarklein replied:
The solution you implemented now causes changes in the API, which should be minimized. I was more aiming for something like:
```py
if isinstance(self, GPUComputePipeline):
    xxx
else:  # GPURenderPipeline
    xxx
```

dskkato replied:
Please wait a moment while I make the corrections.

dskkato replied:
type check was added.

The test code might not be ideal as it borrows from existing code, but for now, I implemented it within the `render_to_texture` function.

rajveermalviya commented:
> > > As implemented in this PR, it seems to work correctly only when the group_id is 0, and otherwise, it crashes.
> > 
> > 
> > I'll have a look too.
> 
> It _nearly_ works, but there is no way (that I know of) to know how many bind groups there are.
> 
> @rajveermalviya this PR implements auto-layouts, it uses `wgpuComputePipelineGetBindGroupLayout()` to get the automatically generated layout objects, but how should we know how many there are? I'm not seeing a way to get this from the shader module object. Does webgpu.h perhaps need something like `wgpuComputePipelineGetBindGroupLayoutCount()`?

IMO this seems like a good feature request for the JS Spec, `getBindGroupLayout` was proposed here - https://github.com/gpuweb/gpuweb/issues/446

> One concern is the current implementation of wgpuComputePipelineGetBindGroupLayout in wgpu-native, which immediately aborts if an out-of-range index is accessed. It would be beneficial if we could either retrieve the number of bind group layouts (like wgpuComputePipelineGetBindGroupLayoutCount), or handle errors at the caller side.

Please file an issue about this over at [gfx-rs/wgpu-native](https://github.com/gfx-rs/wgpu-native)

almarklein commented:
@rajveermalviya thanks for the response.

With the current solution I don't think need `wgpuComputePipelineGetBindGroupLayoutCount()`, because the application-code is (probably) aware of the number of bind groups. So what remains is failing more gracefully when an invalid index is given by accident, which can indeed be solved in the implementation of `wgpuComputePipelineGetBindGroupLayout` and `wgpuRenderPipelineGetBindGroupLayout` in wgpu-native.

dskkato commented on tests/test_wgpu_native_compute_tex.py:
```diff
@@ -526,7 +526,6 @@ def _compute_texture(compute_shader, texture_format, texture_dim, texture_size,
         layout=pipeline_layout,
         compute={"module": cshader, "entry_point": "main"},
     )
-    assert compute_pipeline.get_bind_group_layout(0) is bind_group_layout
```
Unfortunately, it didn't work. I believe the some pointers or handles held by the struct pointed by the  `_internal` should match, but it seems there is now way to verify that.

almarklein replied:
I figured as much ...

almarklein reviewed (approved):
LGTM!

For the moment the panic-on-invalid-index is unfortunate, but not *that* much of an issue to hold off this PR. 

In the mean time, I created a PR to let wgpu-native produce more friendly errors: https://github.com/gfx-rs/wgpu-native/pull/320
